naive implementation of autoactivate mode from conda.el

claytharrison opened the pull request:
I was missing autoactivate mode from conda.el, so I patched it in here. All the code is directly from there but with the necessary name changes to make it work with micromamba. It works mostly well for me except the logic for falling back to the "base" environment for buffers that don't have a detectable env throws an error, since the "base" micromamba env is empty I guess. I'm totally new to elisp and it might take me a while to sort that out, so I'll leave this as a draft for now, but I wanted to at least put it here in case it's something super simple for somebody else.

Commit:
naive implementation of autoactivate mode from conda.el

SqrtMinusOne commented on micromamba.el:
```diff
@@ -194,6 +277,31 @@ The parameters value is an alist as defined by
+           (matches (-filter 'micromamba--env-dir-is-valid possibilities)))
+      (if (> (length matches) 0)
+          (file-name-as-directory (expand-file-name (car matches)))
+        (error "No such conda environment: %s" name)))))
```
I think this function should be replaced by parsing `micromamba env list --json`.

`conda.el` does this because `conda` is written in Python and as such too slow to call. `micromamba` is much, much faster.

This will allow removing a whole bunch of stuff.

SqrtMinusOne commented on micromamba.el:
```diff
@@ -235,5 +343,52 @@ full paths."
+  :global t
+  ;; Forms
+  (if micromamba-env-autoactivate-mode ;; already on, now switching off
+      (advice-add 'switch-to-buffer :after #'micromamba--switch-buffer-auto-activate)
```
Emacs 27.1 introduced a hook called `window-selection-change-functions`, I suggest using that instead of `advice-add`.

SqrtMinusOne commented on micromamba.el:
```diff
@@ -235,5 +343,52 @@ full paths."
+    (if (not (eql env-path nil))
+        (micromamba-activate env-path)
+      (if micromamba-message-on-environment-switch
+          (message "No Conda environment found for <%s>" (buffer-file-name))))))
```
`if` without the alternate clause should be `when`

SqrtMinusOne commented on micromamba.el:
```diff
@@ -48,11 +48,27 @@
+  "Whether to activate the base environment by default if no other is preferred.
+Default nil."
+  :type 'boolean
+  :group 'micromamba)
```
`micromamba` doesn't have `base`, so this doesn't make much sense. I suggest making a variable like `micromamba-fallback-environment`, i.e. an environment that `micromamba-env-autoactivate-mode` activates by default.

With that in mind, parsing `micromamba config list --json` to check for `auto_activate_base` also doesn't seem necessary.

SqrtMinusOne commented on micromamba.el:
```diff
@@ -157,6 +195,51 @@ Returns an alist with the following keys:
+  "Search up the project tree for an `environment.yml` defining a conda env."
+  (let* ((filename (buffer-file-name))
+         (working-dir (if filename
+                          (f-dirname filename)
```
`micromamba` doesn't depend on `f`. I'd prefer implementing that with the standard Elisp filesystem API, unless `f` has something complicated enough to re-implement.

SqrtMinusOne commented on micromamba.el:
```diff
@@ -157,6 +195,51 @@ Returns an alist with the following keys:
+            micromamba-activate-base-by-default
+            (alist-get 'auto_activate_base (micromamba--get-config)))
+           "base"
+         nil)))))
```
`(if <clause> <then> nil)` is equivalent to `(when <clause> <then>)`. There are a bunch of such places, please fix this.

SqrtMinusOne commented on micromamba.el:
```diff
@@ -157,6 +195,51 @@ Returns an alist with the following keys:
+
+(defun micromamba--get-name-from-env-yml (filename) ;; adapted from conda.el
+  "Pull the `name` property out of the YAML file at FILENAME."
+  ;; TODO: find a better way than slurping it in and using a regex...
```
Uhm... TODO? :slightly_smiling_face:

SqrtMinusOne replied:
I'd at least cache the results of this against `file-has-changed-p`. I imagine this would be pretty slow on an HDD.

SqrtMinusOne commented on micromamba.el:
```diff
@@ -89,6 +116,17 @@ Pass ARGS as arguments to the program."
+  (if (not (eq micromamba--config nil))
+      micromamba--config
+    (let ((cfg (micromamba--call-json "config" "list" "--json")))
+      (setq micromamba--config cfg))))
```
I suggested removing that above, but this can be written more concisely as follows:
```lisp
(or micromamba--config
    (setq micromamba-config
          (micromamba--call-json "config" "list" "--json")))
```

SqrtMinusOne commented:
Done a quick review.

Unfortunately, `conda.el` is not a very well-written package to learn elisp :slightly_smiling_face: And workarounds for the slowness of `conda` itself are unnecessary with `micromamba`.

Please address my comments and the CI output (except the rate limit issue) if you want to merge that.

claytharrison committed:
Merge branch 'SqrtMinusOne:master' into autoactivate

claytharrison committed:
Replace (if ... nil) with (when ...)

claytharrison committed:
if w/o alternate -> when

claytharrison committed:
advice-add -> window-selection-change-functions

claytharrison committed:
Use fallback environment instead of base

claytharrison committed:
Fix window-selection-change-functions add

claytharrison committed:
Use micromamba-activate on inferred env (duh)

claytharrison committed:
Remove dependency on f.el

claytharrison committed:
Remove obsolete code

claytharrison committed:
Formatting

claytharrison committed:
Fix deletion frm window-selection-change-functions

claytharrison committed:
Only activate if same env not active

claytharrison committed:
Add caching for yml contents.

claytharrison commented:
Hi! I'm a lot more comfortable with elisp now and I've finally come back to address this stuff. Made it through all your comments and I'm a lot happier with it now. Thanks for the detailed review, hope it's closer to passing muster this time :)

SqrtMinusOne commented on micromamba.el:
```diff
@@ -158,6 +166,51 @@ Returns an alist with the following keys:
+                             `(,filename . ,(micromamba--read-file-into-string filename))))
+            (cdr (assoc filename micromamba--yml-cache)))))
+      (when (string-match "name:[ ]*\\([A-z0-9-_.]+\\)[ ]*$" env-yml-contents)
+          (match-string 1 env-yml-contents)))))
```
A few things here.

First, it's best to avoid `add-to-list` in package code, not least because it performs deep equality comparison on each element behind the scenes. `setf` is the recommended operation:
```lisp
(defvar foo nil)
(setf (alist-get "bar" foo nil nil #'equal) "baz")
(setf (alist-get "bur" foo nil nil #'equal) "brr")
foo ; => (("bur" . "brr") ("bar" . "baz"))
(alist-get "bar" foo nil nil #'equal)   ; => "baz"
```
Unfortunately, `assoc` isn't setf-able.

Then, I don't see much point in storing the whole yaml file in cache.

I'd do it like this:
```lisp
(or (unless (file-has-changed-p filename)
      (cdr (assoc filename micromamba--yml-cache)))
    (<read file normally and store the name in the cache>))
```

Third, it's conventional to wrap the `string-match` and `match-string` in `save-match-data`, so please do that.

SqrtMinusOne replied:
Also, in elisp we have to choose between alists and hashmaps: alists have O(n) random access, but are still faster if there's N < 20 elements. Which seems to be the case here, so it's fine.

SqrtMinusOne commented on micromamba.el:
```diff
@@ -236,5 +294,49 @@ full paths."
+                 #'micromamba--switch-buffer-auto-activate)
+    (setq window-selection-change-functions
+          (delete #'micromamba--switch-buffer-auto-activate
+                  window-selection-change-functions))))
```
It's conventional to use `add-hook` and `delete-hook` for such purposes, please use them.

SqrtMinusOne commented on micromamba.el:
```diff
@@ -158,6 +166,51 @@ Returns an alist with the following keys:
+    (when working-dir
+      (or
+       (micromamba--get-name-from-env-yml (micromamba--find-env-yml working-dir))
+       micromamba-fallback-environment))))
```
This function is going to be called pretty often, so I'd also cache the results of this in a local variable.

Like:

```elisp
(defvar-local micromamba--buffer-env nil
  "The environment to autoactive for the buffer.")
```

Then `setq` will always set the buffer-local value.

SqrtMinusOne replied:
Yeah, and also please check for `file-remote-p` of `(buffer-file-name)` being nil. 

I think it doesn't make much sense to use that on a remote system, but that many filesystem operations slow down TRAMP pretty significantly on a connection like mine.

SqrtMinusOne commented:
Thanks!

It's almost good to go; I've made just a few performance-conscious comments.

claytharrison committed:
Optimize micromamba--get-name-from-env-yml

1. Use setf and alist-get instead of add-to-list
2. Store only env name for buffer in cache
3. Wrap string-match and match-string in save-match-data
4. Pull the string-matching into its own function

claytharrison committed:
Use add-hook/remove-hook instead of setq/delete

claytharrison committed:
Cache micromamba--buffer-env, check file-remote-p

claytharrison commented:
Thanks! That was really educational. I believe everything is addressed, not sure if my logic is the most elegant though.

SqrtMinusOne commented on micromamba.el:
```diff
@@ -158,6 +166,62 @@ Returns an alist with the following keys:
+  "Pull the `name` property out of the YAML file at FILENAME."
+  (when filename
+    (let ((filename (file-truename filename)))
+      (or (unless (file-has-changed-p filename))
+          (cdr (assoc filename micromamba--yml-cache))
```
This definitely should be
```suggestion
      or (unless (file-has-changed-p filename)
            (cdr (assoc filename micromamba--yml-cache)))
```

SqrtMinusOne commented on micromamba.el:
```diff
@@ -158,6 +166,62 @@ Returns an alist with the following keys:
+    (let ((filename (file-truename filename)))
+      (or (unless (file-has-changed-p filename))
+          (cdr (assoc filename micromamba--yml-cache))
+          (setf (alist-get filename micromamba--yml-cache)
```
This won't reliably work because the default test function of `alist-get` is `eq`, which is merely an identity comparison. That is, it returns `t` when the operands are the same object, not when they are two different objects with the same content, which strings might be.

```lisp
(setq foo "bar")

(eq foo foo) ; => t
(eq foo "bar") ; => nil
(eq "bar" "bar"); => nil
(equal foo "bar"); => t
```

So this line needs to be:
```suggestion
          (setf (alist-get filename micromamba--yml-cache nil nil #'equal)
```
This looks a bit hacky, but I haven't found a better way for alists with string keys.

claytharrison committed:
Address changes

SqrtMinusOne commented:
Thanks! Sorry for the late merge. I believe it's good to go.
